janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/AdjacentVertexFilterOptimizerStrategy.java

```diff
@@ -32,73 +32,162 @@
-                    direction = Direction.BOTH;
+        TraversalHelper.getStepsOfClass(TraversalFilterStep.class, traversal)
+            .forEach(originalStep -> {
+                // Check if this filter traversal matches the pattern: _.inV/outV/otherV.is/hasId(x)
+                Traversal.Admin<?, ?> filterTraversal =
+                    (Traversal.Admin<?, ?>) originalStep.getLocalChildren().get(0);
+                List<Step> subSteps = filterTraversal.getSteps();
+
+                OptimizableQueryType type = analyzeSubSteps(subSteps);
+
+                if (type != OptimizableQueryType.NONE) {
+                    replaceStep(traversal, type, originalStep, subSteps);
                 }
-                P predicate = ((IsStep) steps.get(1)).getPredicate();
-                //Check that we have a valid direction and a valid vertex filter predicate
-                if (direction != null && predicate.getBiPredicate() == Compare.eq && predicate.getValue() instanceof Vertex) {
-                    Vertex vertex = (Vertex) predicate.getValue();
-
-                    //Now, check that this step is preceded by VertexStep that returns edges
-                    Step<?, ?> currentStep = originalStep.getPreviousStep();
-                    while (currentStep != EmptyStep.instance()) {
-                        if (!(currentStep instanceof HasStep) && !(currentStep instanceof IdentityStep)) {
-                            break;
-                        } //We can jump over other steps as we move backward
-                        currentStep = currentStep.getPreviousStep();
-                    }
-                    if (currentStep instanceof VertexStep) {
-                        VertexStep vertexStep = (VertexStep) currentStep;
-                        if (vertexStep.returnsEdge()
-                                && (direction == Direction.BOTH || direction.equals(vertexStep.getDirection().opposite()))) {
-                            //Now replace the step with a has condition
-                            TraversalHelper.replaceStep(originalStep,
-                                new HasStep(traversal,
-                                    new HasContainer(ImplicitKey.ADJACENT_ID.name(), P.eq(vertex))),
-                                traversal);
-                        }
-                    }
+            });
+    }
+
+    private void replaceStep(Traversal.Admin<?, ?> traversal, OptimizableQueryType type,
+                             TraversalFilterStep originalStep, List<Step> steps) {
+        // Get the direction in which we filter on the adjacent vertex (or null if not a valid
+        // adjacency filter)
+        Direction direction = parseDirection(steps);
+        P predicate = parsePredicate(type, steps);
+        // Check that we have a valid direction and a valid vertex filter predicate
+        if (direction != null && isValidPredicate(type, predicate)) {
+            if (isPreviousStepValid(originalStep, direction)) {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [These nested if statements could be combined](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=4989535)